### ISSUE TITLE
DPO训练支持channel loss吗？

### ISSUE THREAD

**Issue by kabugge** [modelscope/ms-swift#7246]
如题

**Comment by kabugge** [comment]
@Jintao-Huang 目前不支持吧？可以给我说一下，我试试加上

**Comment by Jintao-Huang** [comment]
是的 还不支持，可以参考 sft的channel loss实现

https://github.com/modelscope/ms-swift/blob/47924ede45f4dd5e5e7b10cc1aef34cbd4cd496b/swift/trainers/trainers.py#L388-L399

**Comment by kabugge** [comment]
> 是的 还不支持，可以参考 sft的channel loss实现
> 
> [ms-swift/swift/trainers/trainers.py](https://github.com/modelscope/ms-swift/blob/47924ede45f4dd5e5e7b10cc1aef34cbd4cd496b/swift/trainers/trainers.py#L388-L399)
> 
> Lines 388 to 399 in [47924ed](/modelscope/ms-swift/commit/47924ede45f4dd5e5e7b10cc1aef34cbd4cd496b)
> 
>  if self.args.enable_channel_loss and channels is not None: 
>      mode = 'train' if self.model.training else 'eval' 
>      metrics = self.custom_metrics[mode] 
>      masks = torch.roll(labels, shifts=-1, dims=-1).view(-1) != -100 
>      if self.template.padding_free: 
>          cu_seqlens = self.get_cu_seqlens(text_position_ids, inputs.get('logits_to_keep')) 
>      else: 
>          cu_seqlens = torch.arange(0, labels.shape[0] + 1) * labels.shape[1] 
>      for i in range(cu_seqlens.shape[0] - 1): 
>          channel = channels[i] 
>          slice_ = slice(cu_seqlens[i], cu_seqlens[i + 1]) 
>          metrics[f'loss_{channel}'].update(outputs.loss[slice_][masks[slice_]])

收到，我把channels = inputs.pop('channel', None)
和trainers.py的388-399行的代码加到了https://github.com/modelscope/ms-swift/blob/47924ede45f4dd5e5e7b10cc1aef34cbd4cd496b/swift/trainers/rlhf_trainer/rlhf_mixin.py#L121这个文件的121行，能运行，不过训练过程似乎跟原来没什么区别，也不知道对不对。。。

**Comment by kabugge** [comment]
这么改的话channels = inputs.pop('channel', None)这行有问题，channels是None

**Comment by kabugge** [comment]
Any update?
